drivers/platform/mbed/mbed_uart.cpp

```diff
@@ -0,0 +1,492 @@
+ */
+static int32_t mbed_uart_set_format(struct uart_init_param *param,
+				    mbed::BufferedSerial *uart)
+{
+	mbed::BufferedSerial::Parity Parity;
+	int data_bits;
+	int stop_bits;
+
+	/* Get the data bits */
+	switch (param->size) {
+	case UART_CS_7:
+		data_bits = 7;
+		break;
+
+	case UART_CS_8:
+		data_bits = 8;
+		break;
+
+	case UART_CS_9:
+		data_bits = 9;
+		break;
+
+	default:
+		/* Invalid data bits. Mbed supports only 7,8 and 9 data bits */
+		return -EINVAL;
+	}
+
+	/* Get the parity */
+	switch (param->parity) {
+	case UART_PAR_NO:
+		Parity = SerialBase::None;
+		break;
+
+	case UART_PAR_EVEN:
+		Parity = SerialBase::Even;
+		break;
+
+	case UART_PAR_ODD:
+		Parity = SerialBase::Odd;
+		break;
+
+	default:
+		/* Invalid parity. Mbed supports only none, odd and even parity */
+		return -EINVAL;
+	}
+
+	/* Get the stop bits */
+	switch (param->stop) {
+	case UART_STOP_1_BIT:
+		stop_bits = 1;
+		break;
+
+	case UART_STOP_2_BIT:
+		stop_bits = 2;
+		break;
+
+	default:
+		/* Invalid stop bits */
+		return -EINVAL;
+	}
+
+	uart->set_format(data_bits, Parity, stop_bits);
+	return SUCCESS;
+}
+
+/**
+ * @brief Initialize the UART communication peripheral.
+ * @param desc[in, out] - The UART descriptor.
+ * @param param[in] - The structure that contains the UART parameters.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t uart_init(struct uart_desc **desc, struct uart_init_param *param)
+{
+	mbed::BufferedSerial *uart;	// Pointer to new BufferedSerial/UART instance
+	platform_usbcdc *usb_cdc_dev;	// Pointer to usb cdc device class instance
+	struct mbed_uart_desc *mbed_uart_desc;	// Pointer to mbed uart descriptor
+	struct uart_desc *uart_desc; 			// UART descriptor
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	// Create the UART description object for the device
+	uart_desc = (struct uart_desc *)calloc(1, sizeof(*uart_desc));
+	if (!uart_desc)
+		return -ENOMEM;
+
+	uart_desc->baud_rate = param->baud_rate;
+
+	if (((struct mbed_uart_init_param *)param->extra)->virtual_com_enable) {
+		// Create a new instance of platform_usbcdc class
+		usb_cdc_dev = new platform_usbcdc(false,
+						  ((struct mbed_uart_init_param *)param->extra)->vendor_id,
+						  ((struct mbed_uart_init_param *)param->extra)->product_id,
+						  ((struct mbed_uart_init_param *)param->extra)->serial_number);
+		if (!usb_cdc_dev) {
+			goto err_usb_cdc_dev;
+		}
+	} else {
+		// Create and configure a new instance of BufferedSerial/UART port
+		uart = new BufferedSerial(
+			(PinName)(((struct mbed_uart_init_param *)param->extra)->uart_tx_pin),
+			(PinName)(((struct mbed_uart_init_param *)param->extra)->uart_rx_pin),
+			(int)param->baud_rate);
+
+		if (!uart)
+			goto err_uart;
+
+		/* Set the UART format */
+		if (mbed_uart_set_format(param, uart) != SUCCESS)
+			goto err_uart;
+	}
+
+	// Create a new mbed descriptor to store new UART instances
+	mbed_uart_desc = (struct mbed_uart_desc *)calloc(1, sizeof(*mbed_uart_desc));
```

Comment:
Can we do this before the previous `if` and include the following `if` in the previous one? If it makes sense it will make for a smoother flow, in my opinion.

Reply:
Yes. I will update the code.